### DISCUSSION TITLE
add autostart option for the volume mixer window

### DISCUSSION THREAD

**Discussion by D4n2021** [File-New-Project/EarTrumpet#1377] [Completed ideas]
After almost a year after I had [already mentioned this](https://github.com/File-New-Project/EarTrumpet/issues/1213), please finally add this, I dislike having to manually open the window _everytime_ we log into Windows.

**Comment by riverar** [comment]
@D4n2021 You want this window to appear, right? And where? Center of screen?
![image](https://github.com/File-New-Project/EarTrumpet/assets/475132/e00de3d0-8e31-40d4-9e71-10a2433af994)

**Reply by D4n2021** [reply]
Exactly! Either center of screen, or, if also possible, that it remembers the position of the window from the last time that it was moved to?

**Reply by riverar** [reply]
The remembering position part is implemented in the dev builds/branch, if you get a moment would be great to get any feedback there. Will look into adding a community setting for opening automatically.

**Reply by TwoCables** [reply]
I'm curious: are you still looking into adding a setting for EarTrumpet's mixer to launch automatically when EarTrumpet is launched? I'm hoping so!

**Reply by riverar** [reply]
@TwoCables I'll add it right now, standby.

**Reply by riverar** [reply]
@TwoCables Done, it's in `dev` and should be automatically built/deployed shortly. (New setting in Community Settings.) Let me know how that works out for you!

**Reply by TwoCables** [reply]
Wow thank you! It's working very well, except for one minor bug that happens when I right-click the icon in the Notification Area. I'll try to describe it:

I have EarTrumpet set to launch delayed on startup by 20 seconds (yep, using Task Manager) so that the mixer is fully populated with everything (otherwise, it's just the Windows volume at the top). When it starts with Windows this way, everything is fine until I right-click the Notification Area icon. Nothing **bad** happens really; all it does is, it makes the mixer the active window instead of giving me a context menu. In addition, I see the context menu for just a split-second when I right-click for the first time. After that, right-clicking the icon always brings up the context menu as expected.

If I don't use Task Scheduler and I let Windows start EarTrumpet for me (set in the Settings app), then this bug doesn't happen, except the mixer doesn't populate anything except the Windows volume at the top. This isn't a bug with the Dev version, it's just the way it has always been for me in the stable release. Interestingly, now that I have the mixer starting up with EarTrumpet, I'm seeing something I've never been able to see before: the mixer starts out with everything populated in it that I want it to have (and continues to gain everything as things finish starting up with Windows that have a volume control), but then a point is reached shortly after startup where it loses all those volume controls leaving just the Windows volume, as though I closed everything. Re-launching the apps doesn't make their volume controls get re-populated in the mixer. I discovered that delaying EarTrumpet's startup makes it work. Apparently, I'm having it start **after** that moment where it loses all of the volume controls. 

Anyway, I love it, and I also love that the mixer really does remember its position. I might not need FancyZones anymore thanks to you. :)

**Reply by riverar** [reply]
@TwoCables Ah yep, I can reproduce this. This is a mixture of a [_foreground rights_](https://learn.microsoft.com/windows/win32/api/winuser/nf-winuser-setforegroundwindow#remarks) constraint in Windows plus some context menu focus confusion in WPF. That is, you didn't launch EarTrumpet directly so when the mixer window opened it couldn't take focus. Then when right-clicking the system tray icon, focus shifted to the shell (explorer.exe), EarTrumpet responded to a system tray message, opened its flyout, then EarTrumpet got focus which confused WPF a bit.

I'll take a look and see if there's a way we can better handle that. Our hands may be a bit tied here, so no promises.

The "blank" issue appears to be related to https://github.com/File-New-Project/EarTrumpet/issues/1305. Do you perhaps have any tasks or programs juggling around audio devices (e.g., disabled/enabled state) at startup?

**Reply by TwoCables** [reply]
Oh hey no it's cool, this is awesome as-is for the foreseeable future. Of course, I still hope it gets fixed.

**Comment by TwoCables** [comment]
> if also possible, that it remembers the position of the window from the last time that it was moved to?

You can use PowerToys for that. https://github.com/File-New-Project/EarTrumpet/discussions/1364

It's not _**precisely**_ what you're talking about, but it's very close.

**Comment by D4n2021** [comment]
> @TwoCables Done, it's in `dev` and should be automatically built/deployed shortly. (New setting in Community Settings.) Let me know how that works out for you!

Omg thanks soooo much! This frees up another hotkey on my keyboard! *_*

**Comment by D4n2021** [comment]
@TwoCables Works AWESOME! Any ETA on when this can move into release build?

**Comment by D4n2021** [comment]
@riverar please implement! :D It works AWESOME!
